### ISSUE TITLE
Initial trade to set probabilities

### ISSUE THREAD

**Issue by leolower** [protofire/gnosis-conditional-exchange#11]

**Comment by fvictorio** [comment]
@cag We implemented the mechanism to make a trade in the created market maker, but we used some hardcoded amount of outcome tokens as a placeholder. Do you have some update on the best way to compute how many tokens have to be bought to set the probabilities?

For example, let's say the user funds the market maker with 100 DAI, and they want to set the initial odds at 75%/25%. How many outcome tokens of the first class it should buy?

**Comment by cag** [comment]
Sorry about the delay. I've done some analysis, and I think I can answer your question now.

---

When the market maker is funded, it immediately splits the funding into outcome tokens and considers that its inventory. When users buy outcome tokens from the market maker, the collateral sent to the outcome token gets split into outcome tokens, and then the market maker sends the requested outcome tokens to the user. Both these steps affect the market maker's inventory.

However, there a convenient expression for estimating an outcome's probability that can be expressed in terms of the market maker's inventory:

    outcomeProbability = numOutcomes ^ (-outcomeTokenHoldings / funding)

Note that this leads naturally to an invariant, as probabilities have to sum up to one:

    sum(outcomeProbabilities) = 1

Given the example, when you start a market maker funded with 100 DAI in funding, it splits that collateral into outcome tokens and begins with an inventory of [100 DAI:A, 100 DAI:B].

You can verify that the market maker's initial estimate of odds for each outcome is:

   2 ^ (-100 / 100) = 50%

---

Let's say a market maker has new outcome probabilities. Then its inventory must have changed to reflect its new probability estimates. We can actually find how the market maker's inventory must have changed:

    newOutcomeProbability                 ( oldOutcomeTokenHoldings - newOutcomeTokenHoldings )
    --------------------- =               ( ------------------------------------------------- )
    oldOutcomeProbability   numOutcomes ^ (                      funding                      )

Or:

    oldOutcomeTokenHoldings - newOutcomeTokenHoldings =
      funding * log(newOutcomeProbability / oldOutcomeProbability) / log(numOutcomes)

If the new odds are 75% A and 25% B, then:

    oldOutcomeTokenHoldingsForA - newOutcomeTokenHoldingsForA =
      100 * log(75% / 50%) / log(2) ~ 58.5

    oldOutcomeTokenHoldingsForB - newOutcomeTokenHoldingsForB =
      100 * log(75% / 50%) / log(2) ~ -100

The market maker _used to_ have 58.5 _more_ DAI:A and 100 _less_ DAI:B. (Be careful of the signs and the order!) For a user setting the odds by buying outcome tokens, this means that after sending `cost` DAI to the market maker which splits that into outcome tokens and adds it to its inventory, and then after the market maker sends back some `outcomeBuyAmount` for each outcome, the market maker will have lost 58.5 DAI:A and gained 100 DAI:B, that is:

    oldOutcomeTokenHoldingsForA - newOutcomeTokenHoldingsForA =
      outcomeBuyAmountForA - cost

    oldOutcomeTokenHoldingsForB - newOutcomeTokenHoldingsForB =
      outcomeBuyAmountForB - cost

To make sure all the `outcomeBuyAmount`s come out nonnegative, just set `cost` to:

    -min((oldOutcomeTokenHoldings - newOutcomeTokenHoldings) for each outcome)

In the example case `cost` is:

    -min(58.5, -100) = 100

And finally that allows us to determine the `outcomeBuyAmounts`: the user should buy [158.5 DAI:A, 0 DAI:B] at a cost of 100 DAI in order to set the odds to 75% A, 25% B.

---

TL;DR:

    netTradeAmountForOutcomes = map (outcome) => funding * log(outcome.targetProbability / outcome.currentProbability) / log(numOutcomes)
    cost = -min(netTradeAmountForOutcomes)
    buyAmounts = netTradeAmountForOutcomes.map(amount => amount + cost)

or

    netTradeAmountForOutcomes =
        [100 * log(75% / 50%) / log(2), 100 * log(25% / 50%) / log(2)] =
        [58.5, -100]
    cost = -min([58.5, -100]) = 100
    buyAmounts = [58.5 + 100, -100 + 100] = [158.5, 0]

**Comment by fvictorio** [comment]
Thanks a lot for the explanation, @cag!

I might be missing something here, because I set up a market maker with 1 DAI of funding, and then made this trade:

```javascript
marketMaker.trade(['1585000000000000000', '0'], 0)
```

and then I check the marginal price:

```javascript
lmsrMarketMaker.calcMarginalPrice(1)
```

and I get `4611596117007846016`. If I understand correctly, you divide this number by 1e18 to get the probability for that outcome, but in this case I don't get neither 25 nor 75. Am I doing something wrong here?

**Comment by cag** [comment]
@fvictorio The fixed-point base I use in those contracts is actually not 1e18, but rather 2**64:

```
>>> 4611596117007846016 / (2**64)
0.24999512643428115
```

It's not completely precise because I truncate the Taylor series in the on-chain computation, but it should be pretty close to the theoretical value.

**Comment by fvictorio** [comment]
Ohh, I see. Thanks!

With respect to the precision, I think that should be more than enough, right? We can just round up the values in the dapp.

**Comment by cag** [comment]
I think so... also it might make sense to add up the probabilities gotten from the contract (so ~25% and ~75% in fixed point), and divide by that "one." Or you could calculate it more precisely off-chain if you know the LMSR state.

In any case, it should be more than enough.
